### ISSUE TITLE
3.2: npm install completes before complaining about missing package.json

### ISSUE THREAD

**Issue by 75lb** [npm/npm#9113]
in npm 3.2, installing a package to a project without a package.json fails (ENOENT package.json).. but it performs all network/downloading first.. 

i work remotely on mobile data plans so pulling a lot of data, only to be told it failed, seems inefficient.. 

is it possible to test for package.json presence *before* tanning the network?

**Comment by iarna** [comment]
A missing `package.json` is a warning, not an error. It shouldn't stop your install from working.

**Comment by iarna** [comment]
Are you seeing it actually stop your install from working?

**Comment by 75lb** [comment]
hi.. yes, it stops the install from working.. move into an empty directory, then: 

```
$ npm i column-layout
/Users/lloyd/Documents/75lb/issues/empty
└── (empty)

npm WARN ENOENT ENOENT, open '/Users/lloyd/Documents/75lb/issues/empty/package.json'
npm WARN EPACKAGEJSON /Users/lloyd/Documents/75lb/issues/empty No description
npm WARN EPACKAGEJSON /Users/lloyd/Documents/75lb/issues/empty No repository field.
npm WARN EPACKAGEJSON /Users/lloyd/Documents/75lb/issues/empty No README data
npm WARN EPACKAGEJSON /Users/lloyd/Documents/75lb/issues/empty No license field.
npm ERR! code 1

$ npm ls
/Users/lloyd/Documents/75lb/issues/empty
└── (empty)

npm ERR! error in /Users/lloyd/Documents/75lb/issues/empty: ENOENT, open '/Users/lloyd/Documents/75lb/issues/empty/package.json'
```

**Comment by 75lb** [comment]
the previous output was from npm 3.2.1

**Comment by saper** [comment]
Same here with npm master (915ed50f90557a97e098d40427dc0dc1e4f6ee66) - so it looks like any successful npm install _package_ in the empty directory results in failure due to `ENOENT`... although it does not, the package is installed.

#9161 is surprising, but this is even more...

**Comment by iarna** [comment]
hmm...

```
11:08 rebecca@Caldina:~/code/npmtest/9113$ node ~/code/npm i column-layout
/Users/rebecca/code/npmtest/9113
└── (empty)

npm WARN ENOENT ENOENT, open '/Users/rebecca/code/npmtest/9113/package.json'
npm WARN EPACKAGEJSON /Users/rebecca/code/npmtest/9113 No description
npm WARN EPACKAGEJSON /Users/rebecca/code/npmtest/9113 No repository field.
npm WARN EPACKAGEJSON /Users/rebecca/code/npmtest/9113 No README data
npm WARN EPACKAGEJSON /Users/rebecca/code/npmtest/9113 No license field.
11:08 rebecca@Caldina:~/code/npmtest/9113$ npm ls
/Users/rebecca/code/npmtest/9113
├── ansi-escape-sequences@2.2.1
├── array-back@1.0.2
├── array-tools@2.0.1
├── collect-all@0.1.0
├── collect-json@1.0.0
├── column-layout@1.0.4
├── command-line-args@1.0.0
├── command-line-usage@1.1.2
├── find-replace@1.0.0
├─┬ object-tools@2.0.3
│ ├── collect-json@0.1.1
│ └── test-value@0.1.0
├── sort-array@0.1.0
├── stream-connect@0.1.0
├── stream-via@0.1.0
├── string-tools@0.1.8
├── test-value@1.0.1
├── typical@2.3.1
└── word-wrap@1.1.0
```

So it seems to (mostly) work for me. It installed all the files correctly. The one thing it _didn't_ do was list the module as being installed at first. 

@75lb Could you try again with `npm i column-layout --loglevel=silly 2> 9113.log` and upload that somewhere for me to look at?

**Comment by saper** [comment]
First, I believe this `Missing package.json` warning is unnecessary or maybe even harmful.

npm indicates failure because exists with 1:

https://gist.github.com/3869151ee7e5a0cc4911

Yes, the files get installed correctly. You get the same error (as well as the `package.json` warning) and the error code 1 if the `node_modules` is already there with that module installed.

**Comment by saper** [comment]
That's interesting - on another box (also running FreeBSD) but with node v0.12.7 I don't get an `NPM Err!` message.

The log comes from FreeBSD box running io.js 3.0

**Comment by 75lb** [comment]
@iarna ok, this time i used npm@3.3.0.. here is the `npm i column-layout --loglevel=silly` output you requested: https://gist.github.com/75lb/15086238cf6f4905a5c7

some more output: 

```
$ node -v
v0.12.7

$ npm ls
/Users/lloyd/Documents/75lb/gists/9113
└── (empty)

npm ERR! error in /Users/lloyd/Documents/75lb/gists/9113: ENOENT, open '/Users/lloyd/Documents/75lb/gists/9113/package.json'

$ tree
.
├── 9113.log
└── node_modules
    ├── ansi-escape-sequences
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── ansi-escape-sequences.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── array-back
    │   ├── LICENSE
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── array-back.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── array-tools
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── bower.json
    │   ├── dist
    │   │   ├── array-tools.js
    │   │   └── array-tools.min.js
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── array-tools.js
    │   ├── package.json
    │   └── test
    │       ├── chaining.js
    │       ├── cli.js
    │       ├── contains.js
    │       ├── exists.js
    │       ├── extract.js
    │       ├── findWhere.js
    │       ├── fixture
    │       │   └── array.json
    │       ├── groupby.js
    │       ├── pick.js
    │       ├── pluck.js
    │       ├── remove.js
    │       ├── spliceWhile.js
    │       ├── unique.js
    │       ├── where.js
    │       └── without.js
    ├── collect-all
    │   ├── LICENSE
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── collect-all.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── collect-json
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── collect-json.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── column-layout
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── example
    │   │   ├── five-columns-width.json
    │   │   ├── five-columns.json
    │   │   ├── no-columns.json
    │   │   ├── one-column-width.json
    │   │   ├── one-column.json
    │   │   ├── table.json
    │   │   ├── two-columns.json
    │   │   └── usage.json
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── column-layout.js
    │   ├── package.json
    │   └── test
    │       ├── fixture
    │       │   ├── empty-column.json
    │       │   └── new-line.json
    │       └── test.js
    ├── command-line-args
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── example
    │   │   ├── alias.js
    │   │   ├── defaultValue.js
    │   │   ├── group.js
    │   │   ├── multiple.js
    │   │   ├── name.js
    │   │   ├── type.js
    │   │   ├── typical.js
    │   │   └── validate.js
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   ├── command-line-args.js
    │   │   ├── definition.js
    │   │   ├── definitions.js
    │   │   └── option.js
    │   ├── package.json
    │   └── test
    │       ├── alias.js
    │       ├── bad-input.js
    │       ├── default-option.js
    │       ├── default-value.js
    │       ├── detect-process-argv.js
    │       ├── getopt.js
    │       ├── grouping.js
    │       ├── missing-def.js
    │       ├── name-alias-mix.js
    │       ├── name-equals.js
    │       ├── type-boolean-multiple.js
    │       ├── type-boolean.js
    │       ├── type-invalid.js
    │       ├── type-none.js
    │       ├── type-number-multiple.js
    │       ├── type-number.js
    │       ├── type-other-multiple.js
    │       ├── type-other.js
    │       ├── type-string.js
    │       └── usage.js
    ├── command-line-usage
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── example
    │   │   ├── my-app.js
    │   │   ├── screens
    │   │   │   └── typical.png
    │   │   ├── typical-formatted.js
    │   │   └── typical.js
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   ├── command-line-usage.js
    │   │   └── usage-options.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── find-replace
    │   ├── LICENSE
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── find-replace.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── object-tools
    │   ├── README.md
    │   ├── bin
    │   │   └── cli.js
    │   ├── bower.json
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── object-tools.js
    │   ├── node_modules
    │   │   ├── collect-json
    │   │   │   ├── LICENSE
    │   │   │   ├── README.md
    │   │   │   ├── bin
    │   │   │   │   └── cli.js
    │   │   │   ├── jsdoc2md
    │   │   │   │   └── README.hbs
    │   │   │   ├── lib
    │   │   │   │   └── collect-json.js
    │   │   │   ├── package.json
    │   │   │   └── test
    │   │   │       └── test.js
    │   │   └── test-value
    │   │       ├── LICENSE
    │   │       ├── README.md
    │   │       ├── jsdoc2md
    │   │       │   └── README.hbs
    │   │       ├── lib
    │   │       │   └── test-value.js
    │   │       ├── package.json
    │   │       └── test
    │   │           └── test.js
    │   ├── package.json
    │   └── test
    │       ├── cli.js
    │       ├── clone.js
    │       ├── exists.js
    │       ├── expression.js
    │       ├── extend.js
    │       ├── extract.js
    │       ├── select.js
    │       ├── where.js
    │       └── without.js
    ├── sort-array
    │   ├── LICENSE
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── sort-array.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── stream-connect
    │   ├── LICENSE
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── stream-connect.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── stream-via
    │   ├── LICENSE
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── stream-via.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── string-tools
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── string-tools.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── test-value
    │   ├── LICENSE
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── test-value.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    ├── typical
    │   ├── README.md
    │   ├── jsdoc2md
    │   │   └── README.hbs
    │   ├── lib
    │   │   └── typical.js
    │   ├── package.json
    │   └── test
    │       └── test.js
    └── word-wrap
        ├── LICENSE
        ├── README.md
        ├── index.js
        └── package.json

91 directories, 190 files
```

**Comment by saper** [comment]
Here's silly log from me: https://gist.github.com/1c9e0ce9bce103670da7

**Comment by saper** [comment]
-- it seems to me that both machines fail now the same way:(

**Comment by iarna** [comment]
Ok, so this is caused by two things, but the biggie is column-layout being required by one of its own deps:

column-layout → command-line-args → command-line-usage → column-layout

This confuses the logical tree generator…

**Comment by saper** [comment]
Nobody promised that dependency graph is a tree :) and the little one?

**Comment by iarna** [comment]
¯\_(ツ)_/¯ I don't honestly know where I was going w/ the little one thing.

And no, no one imagines our dependency graphs are always trees, this one had to with how extraneousness was computed and what we did about it when we were transforming a physical tree (representing the folders on disk) into a logical tree (representing the dependency graph).

The problem was when you had a dependency cycle like that AND the module was extraneous then it (and all of its dependencies) were excluded from the logical tree, which is the basis for `npm ls` output.

**Comment by othiym23** [comment]
Fixed in part by d119ea67804e4bd2ea00025bd1d779cf7ec8ca18, and in part by 02420dc73ae185144267912e461da2effe90341d, included in `npm@3.3.1`. Thanks for all your help and patience, @75lb!
